myia/impl_py.py

```diff
@@ -0,0 +1,160 @@
+"""Implementations for the debug VM."""
+
+
+from typing import Dict, Callable
+from copy import copy
+from myia import primops
+
+
+implementations: Dict[primops.Primitive, Callable] = {}
+
+
+def registrar(pfx, store):
```

Comment:
Should we insist on the prefix? It should be clear that the `add` function is an implementation because it's in the `impl_py` module, so we shouldn't have to enforce the `impl_` prefix everywhere.

Reply:
`@impl` gets the primitive name from the function name by assuming it has the form `impl_{name}`, so checking the prefix is a sanity check of sorts.

Also at some point I was pondering having a second type of implementation, which would take `frame` as an extra argument and could implement primitives like `if`. That would be a different prefix in the same file.

Reply:
Not sure I follow what "sanity check" enforcing `impl_{name}` instead of `{name}` is. It just checks whether I named a function a certain way.

If you have multiple things for a single primitive in the same file, then sure, although I thought the point of this exercise was to separate implementations of certain aspects into different modules after all. That said, can we at least have something readable like `py_add` or `add_implementation`? `impl_add`, and `impl_py` for that matter, aren't very readable names.

Reply:
Getting the primitive name from the function name saves some typing and prevents some copy-paste errors. Adding a prefix is useful since otherwise the implementation of certain primitives that happen to have the same name as a builtin will shadow that builtin, so it needs to be aliased to something else. On the other hand, I suppose it would be nice if you could import `add` from `myia.impl`, use it in a Myia-decorated function, and have it resolve to the primitive.

Reply:
On the other hand, `@implementation(primops.add)` is much more explicit than `@impl` plus magic string matching, which forces me to name my function `impl_bla`. It also enables static checking e.g. `impl_Named` would happily parse right now, but if you type `implementation` you could get a static type error.

Mostly pushing back on this because although this isn't too bad, I think that libraries which add magic in order to save a few keystrokes quickly end up being hard to maintain and difficult to understand for outsiders. Without reading the implementation of `registrar` someone would be surprised to find out that things crash when they change function names, or end up wondering how to register the implementation of their new primitive. All of these small bits of magic quickly introduce significant startup costs to new developers.

Reply:
I don't particularly care about this, so I'll just make it explicit.

I think you are exaggerating the issue, though. The point of this kind of magic isn't *just* to save typing, but also to prevent easy copy/paste errors that would see the primitive changed but not the function name. That has happened to me multiple times, which caused a few really bizarre errors, but that's mostly an issue with rapid development cycles. I agree that the magic needs not be user-facing, though (I was thinking of this mostly for internal use, to be honest).

Reply:
Done.